hkmc2/shared/src/test/mlscript/nofib/mate.mls

```diff
@@ -0,0 +1,652 @@
+  append(
+    (if withPiece
+      then append(
+        showPiece([c, k]),
+        if ((k === King) || (k is Pawn and not(capt || prom))) then Nil else "/" :: showSquare(c, sq)
+      )
+      else Nil
+    ),
+    append(
+      maybe("-" :: Nil, cp => "x" :: append(showPiece(cp), "/" :: Nil), mcp),
+      append(
+        showSquare(c, sq_),
+        maybe(Nil, pp => "(" :: append(showPiece(pp), ")" :: Nil), mpp)
+      )
+    ) 
+  )
+
+fun showMoveInFull(a) = showMove(true, a)
+
+fun showMovesAfter(p_, mifs) = if mifs is
+  Nil then Nil
+  MoveInFull(p, sq, d_) :: mifs and p_ is MoveInFull(p_, sq_, _) then
+    append(
+      nofibStringToList(", "),
+      append(
+        showMove(not(eqTuup2(p, p_)) || not(eqTuup2(sq, sq_)), MoveInFull(p, sq, d_)),
+        showMovesAfter(MoveInFull(p, sq, d_), mifs)
+      )
+    )
+
+fun showMoves(mifs) = if mifs is
+  Nil then throw Error("showMoves")
+  mif :: mifs then append(
+    showMoveInFull(mif),
+    showMovesAfter(mif, mifs)
+  )
+
+fun sift(c, bd, ms, sqs) = if sqs is
+  Nil then ms
+  sq :: sqs and
+    onboard(sq) and pieceAt(bd, sq) is
+      None then sift(c, bd, Move(sq, None, None) :: ms, sqs)
+      Some(p_) and
+        colourOf(p_) === c then sift(c, bd, ms, sqs)
+        else sift(c, bd, Move(sq, Some(p_), None) :: ms, sqs)
+    else sift(c, bd, ms, sqs)
+
+fun moveLine(bd, c, sq, inc, cont) = 
+  fun ml(sq, ms) = 
+    let sq_ = inc(sq)
+    if onboard(sq_) and pieceAt(bd, sq_) is
+      None then ml(sq_, Move(sq_, None, None) :: ms)
+      Some(p_) and
+        not(colourOf(p_) === c) then cont(Move(sq_, Some(p_), None) :: ms)
+        else cont(ms)
+    else cont(ms)
+  ms => ml(sq, ms)
+
+// bishopmoves :: Colour -> Square -> Board -> [Move]
+// bishopmoves c sq bd =
+//   ( moveLine bd c sq (\xy -> case xy of {(x,y) -> (x-1,y+1)}) $ (moveLine bd c sq (\xy -> case xy of {(x,y) -> (x+1,y+1)}) $ (moveLine bd c sq (\xy -> case xy of {(x,y) -> (x-1,y-1)}) $ (moveLine bd c sq (\xy -> case xy of {(x,y) -> (x+1,y-1)}) (\x -> x))))) []
+fun bishopmoves(c, sq, bd) =
+  moveLine(
+    bd,
+    c,
+    sq,
+    xy => if xy is [x, y] then [x - 1, y + 1],
+    moveLine(
+      bd,
+      c,
+      sq,
+      xy => if xy is [x, y] then [x + 1, y + 1],
+      moveLine(
+        bd,
+        c,
+        sq,
+        xy => if xy is [x, y] then [x - 1, y - 1],
+        moveLine(
+          bd,
+          c,
+          sq,
+          xy => if xy is [x, y] then [x + 1, y - 1],
+          x => x
+        )
+      )
+    )
+  )(Nil)
+
+// rookmoves :: Colour -> Square -> Board -> [Move]
+// rookmoves c sq bd =
+//   ( moveLine bd c sq (\xy -> case xy of {(x,y) -> (x-1,y)}) $ (moveLine bd c sq (\xy -> case xy of {(x,y) -> (x+1,y)}) $ (moveLine bd c sq (\xy -> case xy of {(x,y) -> (x,y-1)}) $ (moveLine bd c sq (\xy -> case xy of {(x,y) -> (x,y+1)}) (\x -> x))))) []
+fun rookmoves(c, sq, bd) =
+  moveLine(
+    bd,
+    c,
+    sq,
+    xy => if xy is [x, y] then [x - 1, y],
+    moveLine(
```

Comment:
Why not use `of` to avoid the parenthesis (and achieve something similar to the original's `$`)?

For the record, this works:

```fs
fun f(a)(b) = [a, b]
fun g(a, b) = [a, b]

(f of
    g of
      1
      2
) of 3
//│ = [ [ 1, 2 ], 3 ]
```

Reply:
Thanks for the tip! I have pushed the commit to use `of` to make the code cleaner.